# Dépassement 32 bits
## Pour les entiers

Calcul simple : $1000 * 2000 * 3000$ ...

On s'attend à 6 000 000 000

In [55]:
#include <stdio.h>

int main() {
int x = 1000;
x = x * 2000;
x = x * 3000;

printf("Résultat : %d\n",x);
return 0;}

Résultat : 1705032704


1 milliard seulement ! au lieu de 6 ?!? (et puis cela devrait être un multiple de 1000 au moins, ou un multiple de $1000 * 1000 * 1000$)

rem. : cela aurait pu être pire :

In [58]:
#include <stdio.h>

int main() {
int x = 2000;
x = x * 2000;
x = x * 2000;

printf("Résultat : %d\n",x);
return 0;}

Résultat : -589934592


un résultat négatif (non multiple de 1000) pour le produit de 4 nombres positifs, multiples de 1000 ...

Cependant, le résultat semble cohérent, on peut le prendre dans n'importe quel sens, c'est toujours pareil :

In [63]:
#include <stdio.h>

int main() {
int x = 3000;
x = x * 2000;
x = x * 1000;

printf("Résultat : %d\n",x);
return 0;}

Résultat : 1705032704


on peut même essayer de le décomposer :

In [66]:
#include <stdio.h>

int main() {
int x = 1000;
x = x * 1000;
x = x * 1000;
x = x * 6;

printf("Résultat : %d\n",x);
return 0;}

Résultat : 1705032704


ou

In [84]:
#include <stdio.h>

int main() {
int x = 1000;
x = x * 1000;
x = x * 1000;
x = x + x + x + x + x + x;

printf("Résultat : %d\n",x);
return 0;}

Résultat : 1705032704


et même en comptant 1 à 1 :

In [97]:
#include <stdio.h>

int main() {
int x = 0, i, j, k;
for(i=0;i<1000;i++) {
  for(j=0;j<2000;j++) {
    for(k=0;k<3000;k++) {
      x = x + 1;}}}
printf("Résultat : %d\n\n",x);
printf("toujours %d\n\n",x);
return 0;}

Résultat : 1705032704

toujours 1705032704



Mais attention, 6 000 000 000 - 1 ou 6 000 000 000 + 1 donnent les bons résultats (enfin, si on considère que pour 6 000 000 000 on a le bon résultat) :

In [99]:
#include <stdio.h>

int main() {
int x = 1000;
x = x * 2000;
x = x * 3000;

printf("Résultat : %d %d %d\n",x-1,x,x+1);
return 0;}

Résultat : 1705032703 1705032704 1705032705


## Explication pour les entiers 

Tout cela parce que les nombres sont sur 32 bits ... dès que l'on arrive aux milliards, on retombe sur 0 et on recommence avec 1, 2, 3, ... (plus exactement, après 2^31 [c'est à dire presque 2 milliards] on peut arriver à -2^31, ou si l'on utilise des nombres non signés [entiers naturels] c'est seulement après 2^32 que l'on voit le problème, dans les deux cas après 2^32 [c'est à dire après un peu plus de 4 milliards] on retombe sur 0)

et effectivment 6 000 000 000 - 2^32 = 1705032704

## Pour les nombres à virgule flottante

Pour les nombres à virgule flottante, c'est une autre histoire. Le milliard est dans les limites de ce que ces nombres peuvent représenter :

In [121]:
#include <stdio.h>

int main() {
float x = 1000.0;
x = x * 2000.0;
x = x * 3000.0;

printf("Résultat : %f %e\n",x,x);
return 0;}

Résultat : 6000000000.000000 6.000000e+09


Super, c'est le résultat exact. Mais n'est-ce pas un peu par chance, parce que par contre, x+1, x-1 ... :

In [124]:
#include <stdio.h>

int main() {
float x = 1000.0;
x = x * 2000.0;
x = x * 3000.0;

printf("Résultat : %f %e - %f %e - %f %e\n",x-1,x-1,x,x,x+1,x+1);
return 0;}

Résultat : 6000000000.000000 6.000000e+09 - 6000000000.000000 6.000000e+09 - 6000000000.000000 6.000000e+09


on reste collé à 6 000 000 000.

Et même, si l'on avait voulu décomposer le calcul, cela aurait peut-être été une très mauvaise idée (outre le fait que le calcul de quelques milliards d'additions en flottant prend un peu de temps) :

In [154]:
#include <stdio.h>

int main() {
float x = 0.0;
int i, j, k;
for(i=0;i<1000;i++) {
  for(j=0;j<2000;j++) {
    for(k=0;k<3000;k++) {
      x = x + 1.0;}}}
printf("Résultat : %f %e \n",x,x);
return 0;}

Résultat : 16777216.000000 1.677722e+07 


quoi, seulement 16 millions ?!?

## Explication pour les nombres à virgule flottante

avec les nombres à virgule flottante, les résultats peuvent être plus grands et dépasser qlq milliards [même s'il y a des limites, elles sont beaucoup plus loin], mais ces nombres perdent en précision : pour un nombre de l'ordre du milliard, les 9 chiffres ne sont pas tous significatifs [justes]. DAns le calculs précédent, quand l'écard de magnitude entre x et 1.0 est trop grand, on dépasse la précision maximale et alors 1.0 est négligé : x+1.0 = x ... d'après ce qui précède, c'est donc quand x dépasse 16 millions (soit une vingtaine de bits), il n'y a donc que 7 chiffres significatifs [justes]).